### PR TITLE
Sampler overhaul

### PR DESCRIPTION
**Pull request by Olllom**
Draft for a new iterative sampler API  + data samplers and metropolized replay buffer.

### CODE REVIEW

File: bgflow/bg.py [hunk 1]
```diff
@@ -2,12 +2,12 @@
```
Comment (Olllom) [inline bgflow/bg.py:2]: the only change in this file is that the distinctions between tensors and tuples-of-tensors are now handled by the new `pack_tensor_in_tuple` function.

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -41,7 +41,7 @@ jobs:
       - name: Install dependencies
```
Comment (Olllom) [inline .github/workflows/CI.yml:41]: Here, I am just introducing NetCDF4 as a dependency for the pip-tests. This is an optional dependency to support writing to a kind of trajectory file.

File: bgflow/distribution/sampling/base.py [hunk 3]
```diff
@@ -1,13 +1,25 @@
 import torch
-
+from ...utils.types import unpack_tensor_tuple, pack_tensor_in_list
 
 __all__ = ["Sampler"]
 
 
 class Sampler(torch.nn.Module):
```
Comment (Olllom) [inline bgflow/distribution/sampling/base.py:7]: This now has a `return_hook` to do something with the samples (e.g. reshaping or selecting elements) before they are returned.

File: bgflow/nn/flow/sequential.py [hunk 4]
```diff
@@ -43,8 +43,7 @@ def forward(self, *xs, inverse=False, **kwargs):
             Total volume change as a result of the transformation.
             Corresponds to the log determinant of the Jacobian matrix.
         """
-        n_batch = xs[0].shape[0]
-        dlogp = torch.zeros(n_batch, 1).to(xs[0])
+        dlogp = 0.0
```
Comment (Olllom) [inline bgflow/nn/flow/sequential.py:46]: This removes the assumption that we only have one event shape.

File: bgflow/distribution/sampling/mcmc.py [hunk 5]
```diff
@@ -1,35 +1,257 @@
+"""Markov-Chain Monte Carlo iterative sampling.
```
Comment (Olllom) [inline bgflow/distribution/sampling/mcmc.py:1]: This file now implements the more general Monte Carlo sampling stuff.
I am also leaving the old implementation in for backwards-compatibility. But I want to take it out as soon as we have a working TSF notebook.

**Review by Olllom — COMMENTED** [review]
A few explanations. Everything else should hopefully be clear from the docstrings.

File: bgflow/distribution/sampling/iterative.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+__all__ = ["SamplerState", "default_extract_sample_hook", "IterativeSampler", "SamplerStep"]
+
+
+def default_set_samples_hook(x):
+    """by default, use samples as is"""
+    return x
+
+
+class SamplerState(dict):
```
Comment (Olllom) [inline bgflow/distribution/sampling/iterative.py:39]: Implement as Dataclass?

**Review by jonkhler — CHANGES_REQUESTED** [review]

**Pull request merged**